### PR TITLE
[WIP] Core/Scripts: Full rework Skadi the Ruthless, thx @ManuFE

### PR DESCRIPTION
**Pull request by mik1893**
Implementing missing spells and correcting actual behaviour & movements

### CODE REVIEW

File: src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/boss_skadi.cpp [hunk 1]
```diff
@@ -160,88 +147,52 @@ class boss_skadi : public CreatureScript
...
     {
         boss_skadiAI(Creature* creature) : ScriptedAI(creature), Summons(me)
         {
-            Initialize();
             instance = creature->GetInstanceScript();
-            m_uiMovementTimer = 0;
-            m_uiSummonTimer = 0;
-        }
-
-        void Initialize()
```
Comment (jackpoz) [inline src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/boss_skadi.cpp]: why did you remove Initialize() method ? Now the constructor is not initializing all class fields

File: src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/boss_skadi.cpp [hunk 2]
```diff
@@ -270,196 +215,433 @@ class boss_skadi : public CreatureScript
         void SummonedCreatureDespawn(Creature* summoned) override
         {
             if (summoned->GetEntry() == NPC_GRAUF)
-                m_uiGraufGUID.Clear();
+                m_uiGraufGUID = ObjectGuid::Empty;
```
Comment (jackpoz) [inline src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/boss_skadi.cpp]: Clear() didn't work ?

File: src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/boss_skadi.cpp [hunk 3]
```diff
@@ -270,196 +215,433 @@ class boss_skadi : public CreatureScript
...
+                        case EVENT_SUMMON:
+                            SpawnMobs();
+                            events.ScheduleEvent(EVENT_SUMMON, 32000);
+                            break;
+                        case EVENT_SPAWN_MOVE:
+                            if (!spawns.empty())
+                            {
+                                std::list<Creature*>::const_iterator itr = spawns.begin();
+                                for (itr; itr != spawns.end(); ++itr)
+                                    (*itr)->SetInCombatWithZone();
```
Comment (jackpoz) [inline src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/boss_skadi.cpp]: are you sure it's safe to store Creature*  and dereference them later ?

File: src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/boss_skadi.cpp [hunk 4]
```diff
@@ -270,196 +215,433 @@ class boss_skadi : public CreatureScript
...
+                    switch (step)
+                    {
+                        case 2:
+                            me->RemoveFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_NOT_SELECTABLE | UNIT_FLAG_NON_ATTACKABLE);
+                            me->GetMotionMaster()->MovePoint(0, Location[9]);
+                            if (!triggers.empty())
+                            {
+                                std::list<Creature*>::const_iterator itr = triggers.begin();
+                                for (itr; itr != triggers.end(); ++itr)
+                                    (*itr)->AI()->SetData(0, 0);
```
Comment (jackpoz) [inline src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/boss_skadi.cpp]: Same as the comment above, are you sure it's safe to store a Creature* and dereference it ?

File: src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/boss_skadi.cpp [hunk 5]
```diff
@@ -270,196 +215,433 @@ class boss_skadi : public CreatureScript
...
+                            break;
+                        case EVENT_SUMMON:
+                            SpawnMobs();
+                            events.ScheduleEvent(EVENT_SUMMON, 32000);
+                            break;
+                        case EVENT_SPAWN_MOVE:
+                            if (!spawns.empty())
+                            {
+                                std::list<Creature*>::const_iterator itr = spawns.begin();
+                                for (itr; itr != spawns.end(); ++itr)
```
Comment (jackpoz) [inline src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/boss_skadi.cpp]: this causes a build warning, you can see the full build log on travis https://travis-ci.org/TrinityCore/TrinityCore/builds/66280496#L1729

**Comment by mik1893** [comment]
thx for commenting
script was imported from my repo, that's why there are some unnecessary changes. will rework asap

File: src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/boss_skadi.cpp [hunk 6]
```diff
@@ -274,192 +230,429 @@ class boss_skadi : public CreatureScript
...
+        void SpawnMobs()
         {
-            if (spell->Id == SPELL_HARPOON_DAMAGE)
+            if (firstSummon)
             {
-                m_uiSpellHitCount++;
-                if (m_uiSpellHitCount >= 3)
+                firstSummon = false;
+                for (uint8 i = 0; i < DUNGEON_MODE(5, 6); ++i)
+                    if (Creature* warrior = me->SummonCreature(NPC_YMIRJAR_WARRIOR, SpawnLoc.GetPositionX() + rand() % 5, SpawnLoc.GetPositionY() + rand() % 5, SpawnLoc.GetPositionZ(), 0, TEMPSUMMON_TIMED_DESPAWN_OUT_OF_COMBAT, 30000))
```
Comment (jackpoz) [inline src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/boss_skadi.cpp]: don't use rand(), it triggers static analysis issues

File: src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/boss_skadi.cpp [hunk 7]
```diff
@@ -274,192 +230,429 @@ class boss_skadi : public CreatureScript
...
+                    {
+                        warrior->SetInCombatWithZone();
+                        if (Unit* target = SelectTarget(SELECT_TARGET_RANDOM, 0))
+                            warrior->AI()->AttackStart(target);
+                    }
+                for (uint8 i = 0; i < 4; i++)
```
Comment (jackpoz) [inline src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/boss_skadi.cpp]: use a define as upper bound and use the same define when declaring the array

File: src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/boss_skadi.cpp [hunk 8]
```diff
@@ -274,192 +230,429 @@ class boss_skadi : public CreatureScript
...
+                    me->SummonCreature(NPC_YMIRJAR_WARRIOR, firstWave[i], TEMPSUMMON_CORPSE_TIMED_DESPAWN, 5000);
+                for (uint8 i = 4; i < 6; i++)
```
Comment (jackpoz) [inline src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/boss_skadi.cpp]: use a define as upper bound and use the same define when declaring the array

File: src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/boss_skadi.cpp [hunk 9]
```diff
@@ -274,192 +230,429 @@ class boss_skadi : public CreatureScript
...
-        void JustDied(Unit* /*killer*/) override
+    private:
+        EventMap events;
+        InstanceScript* instance;
+        SummonList Summons;
+        ObjectGuid m_uiGraufGUID;
+        CombatPhase Phase;
+        bool firstSummon;
+        uint8 waveCount;
+        std::list<Creature*> spawns;
```
Comment (jackpoz) [inline src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/boss_skadi.cpp]: never store WorldObject* in class fields , use guids instead

File: src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/boss_skadi.cpp [hunk 10]
```diff
@@ -274,192 +230,429 @@ class boss_skadi : public CreatureScript
...
+                                grauf->AI()->DoAction(DATA_FLYING);
+                            break;
+                        case EVENT_SUMMON:
+                            SpawnMobs();
+                            events.ScheduleEvent(EVENT_SUMMON, 32000);
+                            break;
+                        case EVENT_SPAWN_MOVE:
+                            if (!spawns.empty())
+                            {
+                                for(std::list<Creature*>::const_iterator itr = spawns.begin(); itr != spawns.end(); ++itr)
```
Comment (jackpoz) [inline src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/boss_skadi.cpp]: there are still Creature* pointers even if I wrote at https://github.com/TrinityCore/TrinityCore/pull/14866#discussion_r32251467 about it.

**Comment by jackpoz** [comment]
Please don't squash the commits anymore till the PR is ready to be merged otherwise all previous comments get hidden, even the unfixed ones

File: src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/boss_skadi.cpp [hunk 11]
```diff
@@ -274,192 +230,429 @@ class boss_skadi : public CreatureScript
...
             }
-            for (uint32 i = iStart; i < iEnd; ++i)
-                me->SummonCreature(NPC_TRIGGER, Location[i]);
+        }
+
+    private:
+        EventMap events;
+        uint8 step;
+        ObjectGuid skadiGUID;
+        std::list<Creature*> triggers;
```
Comment (jackpoz) [inline src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/boss_skadi.cpp]: never store WorldObject* in class fields , use guids instead

File: src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/boss_skadi.cpp [hunk 12]
```diff
@@ -165,83 +152,58 @@ class boss_skadi : public CreatureScript
         return GetInstanceAI<boss_skadiAI>(creature);
     }
 
+
     struct boss_skadiAI : public ScriptedAI
     {
         boss_skadiAI(Creature* creature) : ScriptedAI(creature), Summons(me)
```
Comment (MitchesD) [inline src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/boss_skadi.cpp]: use BossAI to get rid of some code

**Comment by mik1893** [comment]
i'm rewriting the whole thing using the new TaskScheduler, i put a WIP
